modules/git-hooks.nix

```diff
@@ -0,0 +1,94 @@
+{ lib, config, pkgs, ... }:
+with lib;
+let
+  cfg = config.git.hooks;
+
+  # These are all the options available for a git hook.
+  hookOption = desc: mkOption {
+    type = types.str;
+    default = "";
+    description = desc;
+  };
+
+  # Only keep all the hooks that have a value set.
+  hooksWithData = filterAttrs (k: v: v != "") config.git.hooks;
+
+  # A collection of all the git hooks in the /bin folder
+  hooksDir = pkgs.buildEnv {
+    name = "git.hooks";
+    paths = mapAttrsToList (k: v: pkgs.writeShellScriptBin k "set -euo pipefail\n\n${v}") hooksWithData;
```

Comment:
Hm, wouldn't hooks allow any interpretable script? like python, etc.

Maybe it's a good idea to limit options here to keep the interface simple.

Could other scripts be easily wrapped?

An alternative wold be to detect if the first line contains a shebang and only use `writeShellScriptBin` if it doesn't. On the other hand there are those substitutions like @shell@ which I'm not familiar with that might break this idea or make its UX awkward when somebody tries to do `#! /usr/bin/env python` (which as I understand it would not be the correct way)

Do you have thoughts on this?

It would be cool to think about an implementation of pre-commit core functionality on top of this PR (out of tree) making use of things like this: https://github.com/pre-commit/pre-commit/blob/master/pre_commit/staged_files_only.py

Reply:
Thinking more about it:

Wouldn't it make sense to have a `stagedFilesOnly` bool option in `hookOption` (or `hookOptionStagedOnly` or similar)?

I'm not sure if I'm having the whole picture here, but at least for some hooks, the stock git hook implementation is actually counter intuitive and almost always needs improvement. For example, why would one want (by default) to apply pre-push or pre-commit hooks on non-staged files?

Reply:
Good point about the bash-only shebangs. My mind is too infected by bash and I forget that there are other scripting languages out there :-p

I didn't follow the `stagedFilesOnly` idea fully, but this seems a bit out of scope for this module. This module should be extensible enough that other modules can extend on top of it.

Reply:
> I didn't follow the stagedFilesOnly idea fully, but this seems a bit out of scope for this module. This module should be extensible enough that other modules can extend on top of it.

I think it is fair to gather feedback on the actual use cases, first, before going down this road. It might eventually pop up again, in the future. The idea refers to one of the uniq seling points of pre-commit which applies anything only on staged files (and temporarily stacks away workdir chages) &mdash; so that a pre-commit / pre-push hook does not fail due to workdir changes which are not subject of the commit / push.